### PR TITLE
Replace D3 Legacy

### LINKED ISSUE #11367: [types-2.0] Parallel Version Structure to Support D3 v4 Standard Bundle Definition and D3 v3
**D3 version 4** can be loaded on a module-by-module basis, but for convenience a _D3 Standard Bundle_ is provided for both module import and vanilla script use.

This D3 Standard bundle is published as a UMD module of name **d3**. **Currently, the definitions for this standard bundle cannot be published** in **DefinitelyTyped/types-2.0** (see closed PR #10453) due to existing dependencies of other libraries on the legacy D3 v3.x definitions (e.g. **plottable**, **nvd3**) contained in the same directory.

This issue is spun off from the original Definitions Request #9936, where it has been tracked so far.

This issue is related to a **functional gap** in the current setup of the **DefinitelyTyped/types-2.0** Branch. The need for a parallel version structure is a known open item as per the TypeScript team, see issue #10711.

This issue carries over the migration issue for the existing D3 version 4 standard bundle defintion file which was tracked under [d3-v4-definitelytyped Issue 116](https://github.com/tomwanzek/d3-v4-definitelytyped/issues/116).

@RyanCavanaugh @andy-ms @mhegazy 

cc @gustavderdrache

With the sincerest hope that it remains a **temporary workaround**, developers who need a definition file representing the D3 version 4 may proceed as follows:

1. Install the module-level definitions for the D3 modules comprising the D3 standard bundle individually, using e.g. `npm install @types/d3-selection --save` and so forth
2. Create your own definition file for the standard bundle by re-exporting the individual module definitions and creating the `d3` global, if required for a vanilla script project (see below for the file content.)
3. Add the hand-crafted definition file to your project compilation context
4. Use

 (a) as a module follows the standard TS module import nomenclature `import * as d3 from 'd3';` (or subsets.
 (b) for a vanilla script, makes the global `d3` available by including `\\\ <reference types='d3' />`

The content of the standard bundle definition file is as follows (`export as namespace d3;` exposes the global):

```ts
// d.ts file

export as namespace d3;

export * from 'd3-array';
export * from 'd3-axis';
export * from 'd3-brush';
export * from 'd3-chord';
export * from 'd3-collection';
export * from 'd3-color';
export * from 'd3-dispatch';
export * from 'd3-drag';
export * from 'd3-dsv';
export * from 'd3-ease';
export * from 'd3-force';
export * from 'd3-format';
export * from 'd3-geo';
export * from 'd3-hierarchy';
export * from 'd3-interpolate';
export * from 'd3-path';
export * from 'd3-polygon';
export * from 'd3-quadtree';
export * from 'd3-queue';
export * from 'd3-random';
export * from 'd3-request';
export * from 'd3-scale';
export * from 'd3-selection';
export * from 'd3-shape';
export * from 'd3-time';
export * from 'd3-time-format';
export * from 'd3-timer';
export * from 'd3-transition';
export * from 'd3-voronoi';
export * from 'd3-zoom';

```

Edit: The correct Definition Request Issue is #9936.

### PR DESCRIPTION
**Pull request by tomwanzek**
This PR closes #11367. It is related to #9936. @andy-ms thanks for making this possible!!!

cc @gustavderdrache @ledragon

The `d3` folder is updated to include a definition file for the **Standard D3 version 4 bundle**.
The definition file allows for using `import`-based scenarios and the use of a `d3` global in vanilla scripts using the standard D3 version 4 bundle as a script. It simply re-exports the members of the constituent modules.

The legacy definitions have been renamed and left in the `d3` folder or reference purposes.

In order to allow definitions of other libraries in DT depending on D3 version 3.5.x to still compile within DefinitelyTyped as well as to have the right dependencies when published to npm `@types`, the following changes where made:
- add a `package.json` stub with the dependency on `@types/d3: "^3.5.36"` (This implies use of the latest published definition file for legacy D3). Dependencies on definitions for other libraries, e.g. `jquery` have so far not been added to the `package.json`. This means that, as before, the type-publisher process will add them in as `latest`.
- the `tsconfig.json` files have been updated as necessary to ensure the definitions resolution respects the legacy definition version of D3 in the `package.json`
- where missing and ascertainable, the version number of the affected library was added to the first line of the comments header as per convention for `DT/types-2.0` to support version number parsing before publishing to npm `@types`
- Each affected definition file has an added `TODO` comment advising of the need to check/update the D3 dependency, if the library gets updated to use D3 version 4.

**No other changes were made to the affected library definitions.**

The following definitions have been updated as per above:
- [x] c3  cc @mcliment
- [x] cal-heatmap cc @RetroChrisB
- [x] d3-tip (possibly, will double check) cc @brspnnggrt
- [x] d3.cloud.layout cc @hansrwindhoff
- [x] d3Kit @morphatic
- [x] dagre-d3 cc @markwongsk
- [x] dc cc @hansrwindhoff @mtraynham
- [x] nvd3 cc @PjMitchell
- [x] plottable

EDIT: Corrected issue reference to 9936 in opening paragraph.

### CODE REVIEW

**Comment by andy-ms** [comment]
Don't see the relation to #9937, did you mean something else?

**Comment by tomwanzek** [comment]
@andy-ms Yes, thanks for catching my typo, I fixed it above.

**Comment by tomwanzek** [comment]
:tada:

**Comment by tomwanzek** [comment]
@andy-ms when you have a chance to merge this PR can you also kindly look at #12272 . It's a tiny fix PR for d3-format (it already passed Travis as well). Thanks!!! most appreciated.

File: c3/index.d.ts [hunk 1]
```diff
@@ -1,10 +1,17 @@
-// Type definitions for C3js
+// Type definitions for C3js v0.4.11
```
Comment (andy-ms) [inline c3/index.d.ts]: No need for a patch version

File: cal-heatmap/tsconfig.json [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
-        "noEmit": true,
-        "forceConsistentCasingInFileNames": true
-    },
-    "files": [
-        "index.d.ts",
-        "cal-heatmap-tests.ts"
-    ]
-}
\ No newline at end of file
+  "compilerOptions": {
```
Comment (andy-ms) [inline cal-heatmap/tsconfig.json]: Use the same indent it had before.

File: d3-tip/package.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
```
Comment (andy-ms) [inline d3-tip/package.json]: Prefer 4 space indents.

File: d3/legacy/plugins/package.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (andy-ms) [inline d3/legacy/plugins/package.json]: This file won't actually do anything right now -- `legacy` will be ignored.

File: nvd3/index.d.ts [hunk 5]
```diff
@@ -1,9 +1,15 @@
-// Type definitions for nvd3 1.8.1
+// Type definitions for nvd3 v1.8.1
```
Comment (andy-ms) [inline nvd3/index.d.ts]: The `v` isn't needed.

File: c3/index.d.ts (hunk 1 above)
Comment (tomwanzek) [inline c3/index.d.ts]: I'll remove it. As an FYI, I tended to keep the patch version in for the d3-xxxxx definitions for ttwo reasons:
(1) It's easier to go back to the underlying library and check for changes in commits after the last validated patch w.r.t. definition  accuracy, and
(2) which goes hand in hand with (1), while a patch release in JS may not break the JS API, there are times when it does break a strongly typed TS API definition.

File: cal-heatmap/tsconfig.json (hunk 2 above)
Comment (tomwanzek) [inline cal-heatmap/tsconfig.json]: Will do.

File: d3-tip/package.json (hunk 3 above)
Comment (tomwanzek) [inline d3-tip/package.json]: Will do.

File: cal-heatmap/tsconfig.json (hunk 2 above)
Comment (tomwanzek) [inline cal-heatmap/tsconfig.json]: FYI. The `.editorconfig` I have in my fork specs out indent=2 for json files? So the changes were simply automatically applied.

**Comment by tomwanzek** [comment]
Pushed the updated changes as per review above. Travis is not starting...yet.

File: cal-heatmap/tsconfig.json (hunk 2 above)
Comment (andy-ms) [inline cal-heatmap/tsconfig.json]: #12280

**Comment by tomwanzek** [comment]
@andy-ms  Seems all in good working order now. Let me know, if you need anything else from my side. Cheers.

**Comment by andy-ms** [comment]
I think you should remove the TODO comments. I don't want people installing types and seeing a TODO comment at the top -- they'll assume it's a work in progress, when it's are actually fine as-is. For example, c3 currently [uses](https://github.com/c3js/c3/blob/master/package.json) d3 version 3.5, so the typings shouldn't be updated yet anyway.

The practice of updating the version number and updating dependencies should be common practices anyway, so we should catch it during code review.

**Comment by tomwanzek** [comment]
@andy-ms as per your request the comments have been removed. I am assuming that Travis will be  finished in appr. 1.5 hrs.

So hopefully, within the next 2hrs we finally have a merge and consecutive publication to the npm @types organization.

The publication of the Standard Bundle definition for version 4 has been a long time coming... Thx in advance for proceeding with this!

**Comment by andy-ms** [comment]
One last thing -- could you eliminate any whitespace-only changes in the diff?

**Comment by tomwanzek** [comment]
@andy-ms While I respect your vigilance, this particular PR process has become rather tedious and protracted for increasingly limited value.

We have worked for **months** to get these definitions published,  with an extreme amount of patience for all the stumbling blocks and delays on the way. I understand the MSFT team, including yourself, are working very hard to provide great tooling. So are others who provide the content.

Tooling alone does not make for a smooth process.

Please confirm that there are no **further** changes required to get this PR published before I consider deleting added lines. Thank you.

cc @RyanCavanaugh

**Comment by andy-ms** [comment]
Yes, this is the last review from me (assuming there are no mistakes made in the next change).

**Comment by tomwanzek** [comment]
Final changes submitted.

**Comment by tomwanzek** [comment]
Final changes passed.

**Comment by tomwanzek** [comment]
Thx!

**Comment by Ciwan1859** [comment]
@tomwanzek, I'm confused! When I [go here](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/d3), is that Type Definitions for D3 v4? or the older v3?

**Comment by eamodio** [comment]
@Ciwan1859 you have to look here https://github.com/DefinitelyTyped/DefinitelyTyped/tree/types-2.0/d3

Only the types-2.0 branch has been updated (as that is what gets pushed to npm)

**Comment by tomwanzek** [comment]
@Ciwan1859  in short as @eamodio just said. With a little background:

All definitions for D3 version 4 are contained in the `DefinitelyTyped/types-2.0` branch as they are written for and with TypeScript 2. So they are contained [here](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/types-2.0/d3). These are the definitions you will acquire when you use TypeScript 2 and acquire definitions in the new 'normal' way by:

`npm install @types/d3 --save`

I.e. `DefinitelyTyped/types-2.0` is currently the default repo which feeds the publication to npm @types.

The definitions in the old `master` branch are still the old (to be phased out) D3 v3.

You can look at #9936 for the history of these changes for D3 and the recently updated [DefinitelyTyped Readme on the main page](https://github.com/DefinitelyTyped/DefinitelyTyped) for the general background.

**Comment by Ciwan1859** [comment]
Thanks guys, I'm still a n00b to Git and GitHub and only just noticed that Branch drop down in the top-left.

Up to this point, I've been manually adding the definitions files. I'll try your command and see if it works!

Thanks again.

**Comment by Ciwan1859** [comment]
@tomwanzek does the `--ambient` flag work on the `npm install @types/d3 --ambient`?

**Comment by tomwanzek** [comment]
@Ciwan1859 the `--ambient` flag you are likely referring to is a legacy of using `typings` to install definitions from DefinitelyTyped (TypeScript 1.8.x).

If you are installing the d3 definitions with `npm install @types/d3 --save` alongside the actual D3 standard bundle `npm install d3 --save`, then you can import from the d3 module in all the usual TypeScript/ES6 ways: `import * as d3 from 'd3';`, `import {select, Selection, arc} from 'd3';` etc.

If you want to use the standard bundle definitions from `@types/d3` **in a vanilla script** (i.e. without a module loader), then you can access the `d3` global by including:

```ts
/// <reference types='d3' />

d3.select('svg');  // Or whatever else your heart desires
...

```
In this latter case, you will have to include the d3 v4 bundle script into your HTML, rather than install/import it.

One note of caution, the vanilla script use is _only_ supported out-of-the-box by the definitions for the standard bundle in `@types/d3`. The individual module definitions (i.e. d3-selection etc.) do not support it out-of-the-box.

Hope this helps.

**Pull request merged**